pkg/syncer/upsync/upsync_process.go

```diff
@@ -0,0 +1,183 @@
+package upsync
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kcp-dev/kcp/pkg/syncer/shared"
+	. "github.com/kcp-dev/kcp/tmc/pkg/logging"
+	"github.com/kcp-dev/logicalcluster/v2"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+const (
+	ResourceVersionAnnotation = "kcp.dev/resource-version"
+)
+
+func (c *Controller) pruneUpstreamResource(ctx context.Context, gvr schema.GroupVersionResource, key string) error {
+	logger := klog.FromContext(ctx)
+	upstreamNamespace, upstreamName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logger.Error(err, "Invalid key")
+		return nil
+	}
+	err = c.upstreamClient.Cluster(c.syncTargetWorkspace).Resource(gvr).Namespace(upstreamNamespace).Delete(ctx, upstreamName, metav1.DeleteOptions{})
+	if err != nil {
+		logger.Error(err, "Could not remove resource upstream")
+		return nil
+	}
+	return nil
+}
+
+func (c *Controller) processDownstreamResource(ctx context.Context, gvr schema.GroupVersionResource, key string) error {
+	logger := klog.FromContext(ctx)
+	downstreamNamespace, downstreamName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logger.Error(err, "Invalid key")
+		return nil
+	}
+
+	logger = logger.WithValues(DownstreamNamespace, downstreamNamespace, DownstreamName, downstreamName)
+
+	var upstreamLocator *shared.NamespaceLocator
+	var locatorExists bool
+	var downstreamObject interface{}
+	// Namespace scoped resource
+	if downstreamNamespace != "" {
+		var nsObj runtime.Object
+		if nsObj, err = c.downstreamNamespaceLister.Get(downstreamNamespace); err != nil {
+			logger.Error(err, "Error getting downstream Namespace from downstream namespace lister", "ns", downstreamNamespace)
```

Comment:
We should not continue here. In case of error, `nsObj` may be `nil`.
We should distinguish between NotFound error and another error.
- In case of NotFound, I assume there's not much we can do, because we can't even find back the upstream namespace to do some cleanup. So just return `nil` to have the reconciliation successful
- In case of another error, just return the error.

Reply:
updated